nengo_loihi/hardware/interface.py

```diff
@@ -690,6 +633,70 @@ def create_io_snip(self):
         self.nengo_io_snip_range = snip_range
 
 
+class HostSocket:
+    recv_retries = 10
+    recv_size = 4096  # python docs recommend small power of 2, e.g. 4096
+    recv_timeout = 0.01
+
+    def __init__(self):
+        self.connected = False
+
+        self.port = np.random.randint(50000, 60000)
+        self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self.socket.setsockopt(socket.SOL_TCP, socket.TCP_NODELAY, 1)
+
+    def connect(self, nxsdk_board):
+        # pause to allow host snip to start and listen for connection
+        time.sleep(0.1)
+
+        host_address = d_get(
+            nxsdk_board,
+            b"ZXhlY3V0b3I==",
+            b"X2hvc3RfY29vcmRpbmF0b3I==",
+            b"aG9zdEFkZHI=",
+        )
+        logger.info("Connecting to host socket at (%s, %s)", host_address, self.port)
+        self.socket.connect((host_address, self.port))
+        self.connected = True
+
+    def close(self):
+        # send -1 to signal host/chip that we're done
+        self.send_all([-1])
+
+        # pause to allow chip to receive -1 signal via host
+        time.sleep(0.1)
+
+        self.socket.close()
+        self.connected = False
+
+    def recv_bytes(self, bytes_expected):
+        data = bytearray([])
+        n_retries = 0
+
+        while len(data) < bytes_expected:
+            ready, _, _ = select([self.socket], [], [], self.recv_timeout)
+            if self.socket in ready:
+                data += bytearray(self.socket.recv(self.recv_size))
+            else:
+                n_retries += 1
+
+        assert len(data) == bytes_expected, "Received (%d) less than expected (%d)" % (
+            len(data),
+            bytes_expected,
+        )
+
+        return np.frombuffer(data, dtype=np.int32)
+
+    def send_all(self, data):
+        msg_bytes = struct.pack("%di" % len(data), *data)
+        msg_len = len(msg_bytes)
+        bytes_sent = 0
+        while bytes_sent < msg_len:
+            _, ready, _ = select([], [self.socket], [])
```

Comment:
Should we have a timeout here, maybe 10 s? (we can make it configurable, too)

Reply:
I don't remember where I read this, but I thought sends always work and can't time out because they just fill a buffer, or something like that... in searching around for a reference on that, I see that the native `socket` library has a [`sendall`](https://docs.python.org/3/library/socket.html#socket.socket.sendall) function that we should likely use instead, as it's implemented in C and likely faster / more reliable.

Reply:
Done in a518db1f